fn/option.go

```diff
@@ -0,0 +1,149 @@
+
+// Some trivially injects a value into an optional context.
+//
+// Some : A -> Option[A].
+func Some[A any](a A) Option[A] {
+	return Option[A]{
+		isSome: true,
+		some:   a,
+	}
+}
+
+// None trivially constructs an empty option
+//
+// None : Option[A].
+func None[A any]() Option[A] {
+	return Option[A]{}
+}
+
+// ElimOption is the universal Option eliminator. It can be used to safely
+// handle all possible values inside the Option by supplying two continuations.
+//
+// ElimOption : (Option[A], () -> B, A -> B) -> B.
+func ElimOption[A, B any](o Option[A], b func() B, f func(A) B) B {
+	if o.isSome {
+		return f(o.some)
+	}
+
+	return b()
+}
+
+// UnwrapOr is used to extract a value from an option, and we supply the default
+// value in the case when the Option is empty.
+//
+// UnwrapOr : (Option[A], A) -> A.
+func (o Option[A]) UnwrapOr(a A) A {
+	if o.isSome {
+		return o.some
+	}
+
+	return a
+}
+
+// WhenSome is used to conditionally perform a side-effecting function that
+// accepts a value of the type that parameterizes the option. If this function
+// performs no side effects, WhenSome is useless.
+//
+// WhenSome : (Option[A], A -> ()) -> ().
+func (o Option[A]) WhenSome(f func(A)) {
+	if o.isSome {
+		f(o.some)
+	}
+}
+
+// IsSome returns true if the Option contains a value
+//
+// IsSome : Option[A] -> bool.
+func (o Option[A]) IsSome() bool {
+	return o.isSome
+}
+
+// IsNone returns true if the Option is empty
+//
+// IsNone : Option[A] -> bool.
+func (o Option[A]) IsNone() bool {
+	return !o.isSome
+}
+
+// FlattenOption joins multiple layers of Options together such that if any of
+// the layers is None, then the joined value is None. Otherwise the innermost
+// Some value is returned.
+//
+// FlattenOption : Option[Option[A]] -> Option[A].
+func FlattenOption[A any](oo Option[Option[A]]) Option[A] {
+	if oo.IsNone() {
+		return None[A]()
+	}
+	if oo.some.IsNone() {
+		return None[A]()
+	}
+
+	return oo.some
+}
+
+// ChainOption transforms a function A -> Option[B] into one that accepts an
+// Option[A] as an argument.
+//
+// ChainOption : (A -> Option[B]) -> Option[A] -> Option[B].
+func ChainOption[A, B any](f func(A) Option[B]) func(Option[A]) Option[B] {
+	return func(o Option[A]) Option[B] {
+		if o.isSome {
+			return f(o.some)
+		}
+
+		return None[B]()
+	}
+}
+
+// MapOption transforms a pure function A -> B into one that will operate
+// inside the Option context.
+//
+// MapOption : (A -> B) -> Option[A] -> Option[B].
+func MapOption[A, B any](f func(A) B) func(Option[A]) Option[B] {
+	return func(o Option[A]) Option[B] {
+		if o.isSome {
+			return Some(f(o.some))
+		}
+
+		return None[B]()
+	}
+}
+
+// LiftA2Option transforms a pure function (A, B) -> C into one that will
+// operate in an Option context. For the returned function, if either of its
+// arguments are None, then the result will be None.
+//
+// LiftA2Option : ((A, B) -> C) -> (Option[A], Option[B]) -> Option[C].
+func LiftA2Option[A, B, C any](
```

Comment:
Shooting in the dark here, but `WrapFn2` could be more descriptive? Similarly we could have `WrapFn1`, `WrapFn3` as needed.

Reply:
The name is taken from [liftA2](https://hackage.haskell.org/package/base-4.19.0.0/docs/Prelude.html#v:liftA2) from Haskell wherein it lifts functions in a pure context into ones that work over other computational contexts. The theoretical grounds for this are the "Applicative Functor" which is where the "A" in "liftA2" comes from. Since Golang doesn't support the truly general version of this function, I postfixed the name with "Option" to indicate which Applicative Functor we were using. I am not fundamentally opposed to renaming it, however, I might suggest that since there is no "truly intuitive" name for a function this general, that using the name it has in other contexts could be useful.

If you agree that there is value to naming things similarly to how they may show up elsewhere, we can leave as is. If you still prefer we call it "Wrap" (wrap how? 😉) we can. Personally I agree with you that liftA2 is not intuitive, but I'd also say that WrapFn2 is not intuitive either. At the end of the day, the intuition for this function comes from its type signature, not its name.

"WrapFn1" is the same thing as [fmap](https://hackage.haskell.org/package/base-4.19.0.0/docs/Prelude.html#v:fmap) which comes from the Functor interface. WrapFn3 has a corollary [liftA3](https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Applicative.html#v:liftA3).

I might suggest the compromise of `LiftFn2Option` to preserve the "lift" aspect which has a bunch of literature on it while also dispensing with the unintuitive "A" part.

Reply:
Looks like the Option stuff got merged as part of #8026